ui/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -0,0 +1,125 @@
+        val processVersion = ProcessVersion(lastDeployAction.processVersionId, ProcessName(details.name), ProcessId(details.id), details.createdBy, details.modelVersion)
+        // TODO: what should be in name?
+        val deployingUser = User(lastDeployAction.user, lastDeployAction.user)
+        val deploymentData = prepareDeploymentData(deployingUser)
+        val deployedScenarioDataTry = resolveGraph(details.json.get).map { resolvedGraph =>
+          DeployedScenarioData(processVersion, deploymentData, GraphProcess(ProcessMarshaller.toJson(resolvedGraph).noSpaces))
+        }
+        Future.fromTry(deployedScenarioDataTry)
+      })
+    } yield dataList
+  }
+
+  def deployProcess(processId: ProcessId, savepointPath: Option[String], comment: Option[String],
+                    performDeploy: (ProcessingType, ProcessVersion, DeploymentData, ProcessDeploymentData, Option[String]) => Future[_])
+                   (implicit user: LoggedUser): Future[ProcessActionEntityData] = {
+    for {
+      processingType <- processRepository.fetchProcessingType(processId)
+      maybeLatestVersion <- processRepository.fetchLatestProcessVersion[DisplayableProcess](processId)
+      latestVersion <- processDataExistOrFail(maybeLatestVersion, processId.value.toString)
+      result <- deployAndSaveProcess(processingType, latestVersion, savepointPath, comment, performDeploy)
+    } yield result
+  }
+
+  private def processDataExistOrFail[T](maybeProcessData: Option[T], processId: String): Future[T] = {
+    maybeProcessData match {
+      case Some(processData) => Future.successful(processData)
+      case None => Future.failed(ProcessNotFoundError(processId))
+    }
+  }
+
+  private def deployAndSaveProcess(processingType: ProcessingType,
+                                   latestVersion: ProcessVersionEntityData,
+                                   savepointPath: Option[String],
+                                   comment: Option[String],
+                                   performDeploy: (ProcessingType, ProcessVersion, DeploymentData, ProcessDeploymentData, Option[String]) => Future[_])(implicit user: LoggedUser): Future[ProcessActionEntityData] = {
+    for {
+      resolvedDeploymentData <- Future.fromTry(resolveDeploymentData(latestVersion.deploymentData))
+      maybeProcessName <- processRepository.fetchProcessName(ProcessId(latestVersion.processId))
+      processName = maybeProcessName.getOrElse(throw new IllegalArgumentException(s"Unknown scenario Id ${latestVersion.processId}"))
+      processVersion = latestVersion.toProcessVersion(processName)
+      deploymentData = prepareDeploymentData(toManagerUser(user))
+      _ <- performDeploy(processingType, processVersion, deploymentData, resolvedDeploymentData, savepointPath)
+      deployedActionData <- actionRepository.markProcessAsDeployed(
+        ProcessId(latestVersion.processId), latestVersion.id, processingType, comment
```

Comment:
I wonder, what will happen when marking process as deployed will end with error? We should cancel deployed scenario?

Reply:
It is only refactor (class was extracted) - this code doesn't changed and I don't want o change anything ;)